File: src/_cpp/ml/NodeExt.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "NodeExt.h"
#include "ml/common.h"
#include "ml/loadxml/xmlProperties.h"
NS_CC_BEGIN

NodeExt::NodeExt()
: _events()
, _actions()
{}

NodeExt::~NodeExt()
{}

bool NodeExt::init()
{
	return true;
}

Node* NodeExt::as_node_pointer()
{
	return dynamic_cast<Node*>(this);
}

Node * NodeExt::getChildByPath( const std::string & path_names )
{
	Node * self = as_node_pointer();
	Node * result = self ? getNodeByPath( self, path_names ) : nullptr;
	return result;
}

Node * NodeExt::getChildByPath( const std::list<int> path_tags )
{
	Node * self = as_node_pointer();
	Node * result = self ? getNodeByTagsPath( self, path_tags ) : nullptr;
	return result;
}

ParamCollection& NodeExt::getParamCollection()
{
	return _params;
}

const ParamCollection& NodeExt::getParamCollection()const
{
	return _params;
}

void NodeExt::load( const std::string & path, const std::string & xmlFile )
{
	std::string file = path;
	if( file.empty() == false && file.back() != '/' )
		file += '/';
	file += xmlFile;
	load( file );
}

void NodeExt::load( const std::string & file )
{
	if( file.empty() )
		return;
	pugi::xml_document doc;
	doc.load_file( file.c_str() );
	auto root = doc.root().first_child();
	
	load( root );
}

void NodeExt::load( const pugi::xml_node & root )
{
	xmlLoader::bookDirectory( this );
	xmlLoader::load( as_node_pointer(), root );
	xmlLoader::unbookDirectory();
}

void NodeExt::runEvent( const std::string & eventname )
{
	auto iter = _events.find( eventname );
	if( iter != _events.end() )
	{
		iter->second.execute( this );
	}
	else
	{
		std::string name = as_node_pointer() ? as_node_pointer()->getName() : "Not node inherited";
		//log_once( "NodeExt[%s]: event with name [%s] not dispatched", name.c_str( ), eventname.c_str( ) );
	}
}

ActionPointer NodeExt::getAction( const std::string & name )
{
	auto iter = _actions.find( name );
	if( iter != _actions.end() )
		return iter->second;
	return nullptr;
}

bool NodeExt::setProperty( int intproperty, const std::string & value )
{
	return false;
}

bool NodeExt::setProperty( const std::string & stringproperty, const std::string & value )
{
	return false;
}

void NodeExt::loadActions( const pugi::xml_node & xmlnode )
{
	FOR_EACHXML( xmlnode, xmlaction )
	{
		std::string actionname = xmlaction.attribute( "name" ).as_string();
		auto action = xmlLoader::load_action( xmlaction );
		assert( action );
		_actions[actionname] = action;
	}
}

void NodeExt::loadEvents( const pugi::xml_node & xmlnode )
{
	FOR_EACHXML( xmlnode, xmllist )
	{
		std::string listname = xmllist.attribute( "name" ).as_string();
		FOR_EACHXML( xmllist, xmlevent )
		{
			auto event = xmlLoader::load_event( xmlevent );
			assert( event );
			_events[listname].push_back( event );
		}
	}
}

void NodeExt::loadParams( const pugi::xml_node & xmlnode )
{
	FOR_EACHXML( xmlnode, xmlparam )
	{
		std::string name = xmlparam.name();
		const auto& attr = xmlparam.attribute( "value" );
		std::string value = attr ? attr.as_string() : xmlparam.text().as_string();
		_params[name] = value;
	}
}

bool NodeExt::loadXmlEntity( const std::string & tag, const pugi::xml_node & xmlnode )
{
	if( tag == xmlLoader::k::xmlTag::ParamCollection )
	{
		loadParams( xmlnode );
	};
	return false;
}


LayerExt::LayerExt() {}
LayerExt::~LayerExt() {}
NodeExt_::NodeExt_() {}
NodeExt_::~NodeExt_() {}
NS_CC_END

File: src/_eu/ml/NodeExt.js
//Define namespace
var EU = EU || {};

EU.NodeExt = function(){

    /** For Test Instance of */
    this.__NodeExt = true;

    /** @type {map<EU.EventsList>} */
    this._events = null;
    /** @type {map<cc.Action>} */
    this._actions = null;
    /** @type {EU.ParamCollection} _params*/
    this._params = null;
    this.initExt = function(){
        this._events = {};
        this._actions = {};
        this._params = new EU.ParamCollection();
    };
    this.as_node_ref= function()
    {
        return (this instanceof cc.Node ? this : null);
    };
    this.getChildByPath_str= function( path_names )
    {
        var self = this.as_node_ref();
        //TODO: EU.Common
        var result = self ? EU.Common.getNodeByPath( self, path_names ) : null;
        return result;
    };
    this.getChildByPath_arr= function(path_tags )
    {
        var self = this.as_node_ref();
        //TODO: EU.Common
        var result = self ? EU.Common.getNodeByTagsPath(self, path_tags) : null;
        return result;
    };
    this.getParamCollection= function()
    {
        return this._params;
    };
    /**
     * @param {String} path
     * @param {String} xmlFile
     */
    this.load_str_n_str= function( path, xmlFile )
    {
        var file = path;
        if( !EU.xmlLoader.stringIsEmpty(file) && file.substr(-1) != '/' )
            file += '/';
        file += xmlFile;
        this.load_str( file );
    };
    this.load_str= function(file )
    {
        if( EU.xmlLoader.stringIsEmpty(file)) return;
        var xmlnode = null;
        EU.pugixml.readXml( file, function(error, data) {
            xmlnode = data;
        }, this);
        var root = xmlnode.firstElementChild;
        if( this.load_xmlnode2 )
            this.load_xmlnode2( root );
        else
            this.load_xmlnode(root);
    };
    this.load_xmlnode= function( root )
    {
        EU.xmlLoader.bookDirectory( this.as_node_ref() );
        EU.xmlLoader.load_node_xml_node( this.as_node_ref(), root, false );
        EU.xmlLoader.unbookDirectory();
    };
    this.runEvent= function( eventname )
    {
        if( eventname in this._events )
            this._events[eventname].execute( this );
    };
    this.getAction= function( name )
    {
        if( name in this._actions )
            return this._actions[name];
        return null;
    };
    this.setProperty_int= function( intproperty, value )
    {
        return false;
    };
    this.setProperty_str= function( stringproperty, value )
    {
        return false;
    };
    /**
     * @param {Element} xmlnode
     */
    this.loadActions= function( xmlnode )
    {
        for(var i=0; i < xmlnode.children.length; i++){
            var xmlaction = xmlnode.children[i];
            var actionname = xmlaction.getAttribute( "name" );
            var action = EU.xmlLoader.load_action_xml_node( xmlaction );
            EU.assert( action );
            this._actions[actionname] = action;
        }
    };
    /**  @param {Element} xmlnode */
    this.loadEvents= function(xmlnode )
    {
        for(var i=0; i < xmlnode.children.length; i++){
            var xmllist = xmlnode.children[i];
            var listname = xmllist.getAttribute( "name" );
            for(var j=0; j < xmllist.children.length; j++){
                var xmlevent = xmllist.children[j];
                var event = EU.xmlLoader.load_event_xml_node( xmlevent );
                EU.assert( event );
                if( !(listname in this._events) )
                    this._events[listname] = new EU.EventsList();
                this._events[listname].push( event );
            }
        }
    };
    /**  @param {Element} xmlnode */
    this.loadParams= function( xmlnode )
    {
        if( !this._params )
        {
            cc.log( "NodeExt: not created this._params. My name is: " + this.as_node_ref().getName());
        }
        for(var i=0; i < xmlnode.children.length; i++){
            var xmlparam = xmlnode.children[i];
            var name = xmlparam.tagName;
            var attr = xmlparam.getAttribute( "value" );
            var value = attr ? attr : xmlparam.textContent;
            this._params.set(name, value);
        }
    };
    /**  @param {String} tag 
     * @param {Element} xmlnode */
    this.loadXmlEntity= function(tag, xmlnode )
    {
        if( tag == EU.xmlLoader.k.ParamCollection )
        {
            this.loadParams( xmlnode );
        }
        return false;
    }
};


EU.LayerExt = cc.Layer.extend({

    __LayerExt: true,

    ctor: function(){
        this._super();
        this.initExt();
    },
    onEnter: function(){
        this._super();
        EU.AudioEngine.playEffect( EU.kSoundLayerExt );
    }
});
EU.NodeExt.call(EU.LayerExt.prototype);

EU.NodeExt_ = cc.Node.extend({
    __NodeExt_: true,

    ctor: function(){
        this._super();
        this.initExt();
    }
});
EU.NodeExt.call(EU.NodeExt_.prototype);
